crates/pera-core/src/dwallet_mpc/network_dkg.rs

```diff
@@ -36,67 +48,248 @@ pub struct DwalletMPCNetworkKeyVersions {
+                key_shares_versions: HashMap::from([(
+                    DWalletMPCNetworkKeyScheme::Secp256k1,
+                    vec![new_key_version],
+                )]),
+                status: DwalletMPCNetworkKeysStatus::Ready(HashSet::from([
+                    DWalletMPCNetworkKeyScheme::Secp256k1,
+                ])),
+            })),
+        }
+    }
+
+    /// Retrieves the *running validator's* decryption key shares for each key scheme
+    /// if they exist in the `key_shares_versions`.
+    ///
+    /// The data is sourced from the epoch's initial system state.
+    /// The returned value is a map where:
+    /// - The key represents the key scheme.
+    /// - The value is a vector of decryption key shares for each network DKG.
+    fn validator_decryption_key_shares(
+        network_mpc_keys: &HashMap<DWalletMPCNetworkKeyScheme, Vec<NetworkDecryptionKeyShares>>,
+        weighted_threshold_access_structure: &WeightedThresholdAccessStructure,
+        decryption_key: ClassGroupsDecryptionKey,
+        party_id: PartyID,
+    ) -> DwalletMPCResult<
+        HashMap<
+            DWalletMPCNetworkKeyScheme,
+            Vec<HashMap<PartyID, <AsyncProtocol as Protocol>::DecryptionKeyShare>>,
+        >,
+    > {
+        network_mpc_keys
+            .into_iter()
+            .map(|(key_scheme, network_mpc_key_versions)| {
+                let shares = Self::get_decryption_key_shares_from_public_output(
+                    &network_mpc_key_versions,
+                    party_id,
+                    weighted_threshold_access_structure,
+                    decryption_key,
+                )?;
+
+                Ok((*key_scheme, shares))
+            })
+            .collect::<DwalletMPCResult<HashMap<_, _>>>()
+    }
+
+    /// The network DKG public output holds the all decryption key shares, the shares are encrypted.
+    /// This function decrypts the shares of the current validator with the decryption (Class-Groups) key
+    /// and returns the decryption key shares.
+    fn get_decryption_key_shares_from_public_output(
+        shares: &Vec<NetworkDecryptionKeyShares>,
+        party_id: PartyID,
+        weighted_threshold_access_structure: &WeightedThresholdAccessStructure,
+        decryption_key: ClassGroupsDecryptionKey,
+    ) -> DwalletMPCResult<Vec<HashMap<PartyID, <AsyncProtocol as Protocol>::DecryptionKeyShare>>>
+    {
+        shares
+            .iter()
+            .map(|share| {
+                let dkg_public_output = <Secp256k1Party as CreatePublicOutput>::new(
+                    &share.encryption_key,
+                    &share.reconstructed_commitments_to_sharing,
+                    &share.current_epoch_encryptions_of_shares_per_crt_prime,
+                )?;
+
+                let secret_share = dkg_public_output.default_decryption_key_shares::<
+                    SECP256K1_SCALAR_LIMBS,
+                    FUNDAMENTAL_DISCRIMINANT_LIMBS,
+                    secp256k1::GroupElement,
```

Comment:
This of course would make this code non-generic, and would make it hard to integrate `ristretto` support. Generics are there to be used, not ignored.

Reply:
Write your code using the same generics as I have, and then instantiate it once for `secp256k1` and once for `ristretto`